dvc/external_repo.py

```diff
@@ -65,66 +68,206 @@ def clean_repos():
         _remove(path)
 
 
-class ExternalRepo(Repo):
-    def __init__(self, root_dir, url):
-        super().__init__(root_dir)
-        self.url = url
-        self._set_cache_dir()
-        self._fix_upstream()
+class BaseExternalRepo:
+    _tree_rev = None
+    _local_cache = None
+
+    @contextmanager
+    def use_cache(self, cache):
+        """Use specified cache instead of erepo tmpdir cache."""
+        self._local_cache = cache
+        if hasattr(self, "cache"):
+            save_cache = self.cache.local
+            self.cache.local = cache
+        # make cache aware of our repo tree
+        with cache.erepo_tree(self.tree):
+            yield
+        if hasattr(self, "cache"):
+            self.cache.local = save_cache
+        self._local_cache = None
```

Comment:
This context manager lets us use an existing local cache instead of the default tmpdir erepo cache. Before we would just edit `erepo.cache.local.cache_dir` to make downloads go where we wanted them to, but we were not using other cache settings like the proper link types.

combined w/the `Remote.erepo_tree` context manager, we can now use the correct local cache instance, and also make sure that local cache instance is aware of the erepo git tree